src/Storages/MergeTree/AsyncBlockIDsCache.cpp

```diff
@@ -0,0 +1,101 @@
+#include <Storages/MergeTree/AsyncBlockIDsCache.h>
+#include <Storages/StorageReplicatedMergeTree.h>
+
+#include <chrono>
+#include <mutex>
+
+namespace DB
+{
+
+std::vector<String> AsyncBlockIDsCache::getChildren()
+{
+    auto zookeeper = storage.getZooKeeper();
+
+    auto watch_callback = [&](const Coordination::WatchResponse &)
+    {
+        auto now = std::chrono::steady_clock::now();
+        if (now - last_updatetime < update_min_interval)
+        {
+            std::chrono::milliseconds sleep_time = std::chrono::duration_cast<std::chrono::milliseconds>(update_min_interval - (now - last_updatetime));
+            task->scheduleAfter(sleep_time.count());
+        }
+        else
+            task->schedule();
+    };
+    std::vector<String> children;
+    Coordination::Stat stat;
+    zookeeper->tryGetChildrenWatch(path, children, &stat, watch_callback);
+    return children;
+}
+
+void AsyncBlockIDsCache::update()
+try
+{
+    std::vector<String> paths = getChildren();
+    Cache cache;
+    for (String & p : paths)
+    {
+        cache.insert(std::move(p));
+    }
+    LOG_TRACE(log, "Updating async block succeed. {} block ids has been updated.", cache.size());
+    {
+        std::lock_guard lock(mu);
+        cache_ptr = std::make_shared<Cache>(std::move(cache));
+        ++version;
+    }
+    cv.notify_all();
+    last_updatetime = std::chrono::steady_clock::now();
+}
+catch (...)
+{
+    LOG_INFO(log, "Updating async block ids cache failed. Reason: {}", getCurrentExceptionMessage(false));
+    task->scheduleAfter(update_min_interval.count());
+}
+
+AsyncBlockIDsCache::AsyncBlockIDsCache(StorageReplicatedMergeTree & storage_)
+    : storage(storage_),
+    update_min_interval(storage.getSettings()->async_block_ids_cache_min_update_interval_ms),
+    path(storage.zookeeper_path + "/async_blocks"),
+    log_name(storage.getStorageID().getFullTableName() + " (AsyncBlockIDsCache)"),
+    log(&Poco::Logger::get(log_name))
+{
+    task = storage.getContext()->getSchedulePool().createTask(log_name, [this]{ update(); });
+}
+
+void AsyncBlockIDsCache::start()
+{
+    if (storage.getSettings()->use_async_block_ids_cache)
+        task->activateAndSchedule();
+}
+
+/// Caller will keep the version of last call. When the caller calls again, it will wait util gets a newer version.
+Strings AsyncBlockIDsCache::detectConflicts(const Strings & paths, UInt64 & last_version)
+{
+    if (!storage.getSettings()->use_async_block_ids_cache)
+        return {};
+
+    std::unique_lock lk(mu);
+    cv.wait_for(lk, update_min_interval, [&]{return last_version == 0 || version != last_version;});
```

Comment:
Why do we wait here with a timeout? Will we just use an old version if the update is not ready? I don't think I fully understand this logic and it contradicts comment before the function. Maybe add comments in header file with simple description of intentions?

Reply:
Why do we need special logic for last_version == 0? We do not want to use this cache for the first call?

Reply:
In some cases, for example the zookeeper connection is unstable so we cannot update the cache, I would like to avoid blocking in this function forever. 
However, that is a paradox that we actually need zookeeper to commit this write. Any suggestion?

The first call check is not necessary. I will change it.

Reply:
If the only real problem (besides just simple timeout due to overload or idling for whatever reason) that can happen is zk connection lost, then I would suggest to block, because zk is required anyway. But I'm not sure if this is true. Let's consult about this tricky spot.

@alesapin, what do you think about the current updating policy for this cache from zk (every 100ms)? Especially, about failures, what should we do if it fails for any reason?

Reply:
Actually even more simple question. Do we really need to block at all? This cache is anyway async. (Maybe I just didn't understand the basic idea about how it should work.)

Reply:
Two explaination:
1. Why do we try to get newer cache everytime? Because when retry happens, it has checked this cache but unlucky has not got the newest block ids, so we keep the old version and tell the cache: give me a newer version. If we retry too fast, it's possible that the cache is not udpated in time.
2. Why timeout? Now I can recall it! When we ask the cache and resolve some conflicts, we are going to ask the cache again in the retry. By this time, the cache is not necessarily newer than before, because it's possible that the cache has been the newest last time! To avoid deadlock, the timeout is introduced.

Reply:
But it sounds to me like we are adding unnecessary latency to a query during the commit (for yet another ZK round-trip, but at most 100ms). I was thinking that this cache should be async, meaning no waiting whatsoever. Btw, if we even wait, it does "not guarantee" any kind of consistency, only "better chances" and we have to pay for it with increased latency. Maybe we can make it completely async?

Reply:
Let's see the write workloads:
- **writes without duplicates**. It will not increase latency, because the first access of cache is always non-block ---- the request version is zero and cache version is non-zero. This cases are of the most.
- **writes with duplicates**
  - detected by cache. Lucky that cache filtered some duplicated inserts, and during retry, we are possible to wait for more 100ms. Well it does not make sense, so in the newest commit, I reset the request version to zero if cache detects duplication.
  - detected by zookeeper, unlucky that the cache doesn't help us detect duplication, **and zookeeper always return us one conflict hash id, rather than all of them**, that's why we need cache. In this case, we should wait a little while for a newer cache, that's why I add this logic. The wait is necessary because **if we don't refresh cache, we are sure to fail again**

Reply:
Thanks for the explanation! Now it makes sense to me.

Reply:
(maybe copy-paste this explanation to somewhere in code)

Reply:
good, added